File: tree/108.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


#recursive solution
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> TreeNode:
        
        if not nums:
            return None
        pivot = len(nums) // 2
        
        root = TreeNode(nums[pivot])
        root.left = self.sortedArrayToBST(nums[:pivot])
        root.right = self.sortedArrayToBST(nums[pivot+1:])
        
        return root
    

#non-ercursive solution
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> TreeNode:
        
        if not nums:
            return None
        
        pivot = (len(nums)-1) // 2 
        
        l = []
        root = TreeNode(nums[pivot])
        l.append([root, 0, pivot-1])
        l.append([root, pivot+1, len(nums)-1])
        
        while l:
            cur, left, right = l.pop()
            
            if left <= right:
                pivot = left + (right - left) // 2 #注意与(right - left) // 2的区别 
                node = TreeNode(nums[pivot])
                if nums[pivot] <= cur.val:
                    cur.left = node
                else:
                    cur.right = node
                    
                l.append([node, left, pivot-1])
                l.append([node, pivot+1, right])
            
        return root

File: tree/108.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        return recur(nums, 0, nums.size()-1);
    }
private:
    TreeNode* recur(std::vector<int>& nums, int begin, int end) {
        if (begin > end) {
            return nullptr;
        }
        int mid = (begin + end) / 2;
        TreeNode* root = new TreeNode{nums[mid]};
        root->left = recur(nums, begin, mid-1);
        root->right = recur(nums, mid+1, end);
        return root;
    }
};
